make/devices.cmake

```diff
@@ -0,0 +1,754 @@
+    list(APPEND RR_PATCH_DEPS ${DEVICE_MERGED_FILE_TARGET})
+
+    # Generate the "real" rr_graph.xml from the default rr_graph.xml file
+    string(CONFIGURE ${RR_PATCH_CMD} RR_PATCH_CMD_FOR_TARGET)
+    separate_arguments(
+      RR_PATCH_CMD_FOR_TARGET_LIST UNIX_COMMAND ${RR_PATCH_CMD_FOR_TARGET}
+    )
+    add_custom_command(
+      OUTPUT ${OUT_RRXML_REAL}
+      DEPENDS ${RR_PATCH_DEPS} ${RR_PATCH_TOOL} ${OUT_RRXML_VIRT}
+      COMMAND ${RR_PATCH_CMD_FOR_TARGET_LIST}
+      VERBATIM
+    )
+
+    make_file_target(FILE ${OUT_RRXML_REAL_FILENAME} GENERATED)
+
+    set_target_properties(
+      ${DEFINE_DEVICE_DEVICE}
+      PROPERTIES
+        ${PACKAGE}_OUT_RRXML_REAL
+        ${CMAKE_CURRENT_SOURCE_DIR}/${OUT_RRXML_REAL_FILENAME}
+    )
+
+    add_custom_target(
+      ${DEFINE_DEVICE_ARCH}_${DEFINE_DEVICE_DEVICE}_${PACKAGE}_rrxml_real
+      DEPENDS ${OUT_RRXML_REAL}
+    )
+  endforeach()
+endfunction()
+
+function(DEFINE_BOARD)
+  # ~~~
+  # DEFINE_BOARD(
+  #   BOARD <board>
+  #   DEVICE <device>
+  #   PACKAGE <package>
+  #   PROG_TOOL <prog_tool>
+  #   [PROG_CMD <command to use PROG_TOOL>
+  #   )
+  # ~~~
+  #
+  # Defines a target board for a project.  The listed device and package must
+  # have been defined using DEFINE_DEVICE.
+  #
+  # PROG_TOOL should be an executable that will program a bitstream to the
+  # specified board. PROG_CMD is an optional command string.  If PROG_CMD is not
+  # provided, PROG_CMD will simply be ${PROG_TOOL}.
+  #
+  set(options)
+  set(oneValueArgs BOARD DEVICE PACKAGE PROG_TOOL PROG_CMD)
+  set(multiValueArgs)
+  cmake_parse_arguments(
+    DEFINE_BOARD
+    "${options}"
+    "${oneValueArgs}"
+    "${multiValueArgs}"
+    ${ARGN}
+  )
+
+  add_custom_target(${DEFINE_BOARD_BOARD})
+  foreach(ARG DEVICE PACKAGE PROG_TOOL PROG_CMD)
+    set_target_properties(
+      ${DEFINE_BOARD_BOARD}
+      PROPERTIES ${ARG} "${DEFINE_BOARD_${ARG}}"
+    )
+  endforeach()
+endfunction()
+
+function(ADD_OUTPUT_TO_FPGA_TARGET name property file)
+  make_file_target(FILE ${file} GENERATED)
+  set_target_properties(${name} PROPERTIES ${property} ${file})
+endfunction()
+
+function(ADD_FPGA_TARGET)
+  # ~~~
+  # ADD_FPGA_TARGET(
+  #   NAME <name>
+  #   TOP <top>
+  #   BOARD <board>
+  #   SOURCES <source list>
+  #   TESTBENCH_SOURCES <testbench source list>
+  #   [INPUT_IO_FILE <input_io_file>]
+  #   [EXPLICIT_MAKE_FILE_TARGET]
+  #   )
+  # ~~~
+  #
+  # ADD_FPGA_TARGET defines a FPGA build targetting a specific board.  By
+  # default input files (SOURCES, TESTBENCH_SOURCES, INPUT_IO_FILE) will be
+  # implicitly passed to MAKE_FILE_TARGET.  If EXPLICIT_MAKE_FILE_TARGET is
+  # supplied, this behavior is supressed.
+  #
+  # The SOURCES file list will be used to synthesize the FPGA images.
+  # INPUT_IO_FILE is required to define an io map. TESTBENCH_SOURCES will be
+  # used to run test benches.
+  #
+  # Targets generated:
+  #
+  # * <name>_eblif - Generate eblif file.
```

Comment:
Should these be <name>_<board>_ so we can build the same project for multiple boards? like we can with `make BOARD=icevision` for example

Reply:
I think the answer applies to the answer about add_subdirectory above, explicit of implicit.  If you support multiple boards, you would call add_fpga_target for each board.  I think only the synthesis and pack is reusable, the place and route will have to be regenerated.  If we think there is a lot of value in sharing the synthesis and pack output, I can refactor to support that.

Reply:
FYI, for each design board support is implied by the presence of a `.pcf` file for a board.

Reply:
Explicit over implicit here.  For example, you might want to have multiple .pcf for a particular board.

Reply:
By reusable do you mean the commands to run, or the output from the commands?

I really like that the current system forces you to name the PCF file exactly the same in each directory. It means we don't end up with people calling the same boards different names....